academic_observatory_workflows/workflows/data_quality_workflow.py

```diff
@@ -0,0 +1,437 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Author: Alex Massen-Hane
+
+
+from __future__ import annotations
+
+import hashlib
+import logging
+import os
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import pendulum
+from airflow import DAG
+from airflow.operators.python import PythonOperator
+from airflow.sensors.external_task import ExternalTaskSensor
+from airflow.utils.task_group import TaskGroup
+from google.cloud import bigquery
+from google.cloud.bigquery import Table as BQTable
+
+from academic_observatory_workflows.config import schema_folder as default_schema_folder, Tag
+from observatory.platform.bigquery import (
+    bq_table_id_parts,
+    bq_load_from_memory,
+    bq_create_dataset,
+    bq_run_query,
+    bq_table_exists,
+    bq_select_columns,
+    bq_get_table,
+    bq_select_table_shard_dates,
+    bq_table_id as make_bq_table_id,
+)
+from observatory.platform.observatory_config import CloudWorkspace
+from observatory.platform.workflows.workflow import Workflow, set_task_state, Release
+
+
+@dataclass
+class Table:
+    project_id: str
+    dataset_id: str
+    table_id: str
+    is_sharded: bool
+    fields: List[str]
+
+    """Create a metadata class for tables to be processed by the DQC Workflow.
+
+    :param project_id: The Google project_id of where the tables are located.
+    :param dataset_id: The dataset that the table is under.
+    :param table_id: The name of the table (not the full qualifed table name).
+    :param is_sharded: True if the table is shared or not.
+    :param fields: Location of where the primary key is located in the table e.g. doi, could be multiple different identifiers.
+    """
+
+    @property
+    def full_table_id(self):
+        return f"{self.project_id}.{self.dataset_id}.{self.table_id}"
+
+
+def make_tables(project_id: str, dataset_id: str, tables: Dict) -> List[Table]:
+    """Create a list table objects for the DQC Workflow to process.
+
+    :param project_id: The Google project ID that the tables are stored in.
+    :param dataset_id: Dataset of where the tables are stored.
+    :param tables: Dictionary of metadata of the given tables, including if the tables are sharded and the primary fields.
+    :return: A list of table objects for the workflow to process.
+    """
+
+    return [
+        Table(
+            project_id=project_id,
+            dataset_id=dataset_id,
+            table_id=table["table_id"],
+            is_sharded=table["is_sharded"],
+            fields=table["fields"],
+        )
+        for table in tables
+    ]
+
+
+class DataQualityWorkflow(Workflow):
+    def __init__(
+        self,
+        *,
+        dag_id: str,
```

Comment:
If we run this on all of the past BigQuery sharded tables, it will pull many of them out of long term storage, which would suddenly increase our BigQuery storage costs if we run it on everything. So we should add a parameter where you can specify the maximum number of past shards to analyse.